src/aqua/installation-spell/src/aqua/deal_spell.aqua

```diff
@@ -60,15 +61,43 @@ func deal_install(ipfs: Multiaddr, worker_def_cid: CID, deal_id: Address):
         log(["Worker installation finished with status", status.str])
 
         if status.str == "install finished":
-            subnet_id, cre_errs <- createSubnetwork(deal_id)
-            log(["Subnet created", subnet_id])
-            -- subnet_id will be 'nil' if error happened on 'createSubnetwork'
+            subnet_id: ?string
+
+            now <- Peer.now_sec()
+            created <- Spell.get_u32("subnet_created")
+            registered <- Spell.get_u32("worker_registered")
+            if created.success:
+                if registered.success:
+                    if now - created.value > DISCOVERY_TIMEOUT_SEC:
+                        -- Create subnet only if DISCOVERY_TIMEOUT_SEC has passed
+                        subnet_id, cre_errs <- createSubnetwork(deal_id)
+                        if subnet_id != nil:
+                            log(["Subnet created", subnet_id])
+                        else:
+                            log(["error creating subnetwork", cre_errs])
+
+                    else:
+                        if now - registered > DISCOVERY_TIMEOUT_SEC:
+                            sid <- Spell.get_string("subnet_id")
+                            if sid.success:
+                                -- Set subnet_id only if DISCOVERY_TIMEOUT_SEC has passed
+                                -- Intention is to rate limit 'registerWorker' call
+                                subnet_id <<- sid.value
+                            else:
+                                log(["error reading subnet_id", sid])
+                else:
+                    log(["error reading worker_registered", registered])
+            else:
+                log(["error reading subnet_created", created])
+
+            -- subnet_id will be 'nil' if DISCOVERY_TIMEOUT_SEC has not yet passed
             if subnet_id != nil:
```

Comment:
Doesn't it mean that the `subnet_id` always will be `nil`, since `"subnet_created"` and `"worker_registered"` will be set only if `subnet_id` is not `nil` and `subnet_id` is set when they're set.

Reply:
`subnet_id` is set if `now - created.value > DISCOVERY_TIMEOUT_SEC`. For example, at the beginning, `created.value` will be `0`.

Reply:
I guess confusion comes from `created.success` vs `created.absent`

Reply:
Oh, I see. Maybe it makes sense to rewrite this bit to make the logic less obscure, wdyt? Or add a comment or smth. I'm not sure I'll remember this the next time I see this code.